### DISCUSSION TITLE
⚫️ Dark Mode for UI !

### DISCUSSION THREAD

**Discussion by YunYouJun** [element-plus/element-plus#3754] [Ideas]
Our theme variable system has undergone a series of transformations, and it's time to start the **dark mode**.

Our designer @huangzhongzhong is working on the color design. Once completed, the dark mode will follow up immediately.

If you have any suggestions, you are welcome to submit them here.

---

Development team members use their spare time to develop, so please be forgiven for the progress.
If you are interested in participating in the design you can contact @huangzhongzhong.

**Comment by jinmarcus** [comment]
有没有可能把主题var的那些配色也通过`el-config-provider`组件来配置和改变？这样可以更细粒度的通过状态更新来很轻松平滑的切换主题，以及控制单个组件的颜色等状态值

**Reply by YunYouJun** [reply]
本应当是可能的，但是目前的 scss 需要通过配置的色彩去预生成不用的色阶来给组件用，所以切换主题色想要完全在运行时跑需要引入 JS 的色彩库来生成色阶，这部分的引入代价需要考虑。
以及色彩是否有通用的第三方选择还是集成在 element-plus 中。

**Reply by tolking** [reply]
直接在 `el-config-provider` 中使用scss生成的css变量应该更好实现些。然后提供一个网页生成自定义的颜色配置

el-config-provider 只需要实现内连css变量

**Reply by YunYouJun** [reply]
如果只是 CSS 变量的话，理论上直接覆盖引入即可，似乎无需通过 `el-config-provide`。👀

**Reply by tolking** [reply]
基于 `el-config-provide` 实现只是增加了一种使用方式，而且能够带来更细粒度地改变组件颜色。例如：全局使用浅色主题配置，侧边使用深色主题配置

**Comment by ExtReMLapin** [comment]
For the guys who really can't wait, I did this : https://github.com/ExtReMLapin/element-plus-darktheme

Include the CSS in your page and it should work

**Comment by emojiiii** [comment]
个人感觉目前element-plus的scss设置不是很合理，比如在更改按钮的type时，对应的样式直接写定了颜色，但是我想如果用的是:root根节点的颜色，后面再支持黑暗模式会不会简单很多，而且会更加灵活，因为type取的颜色是继承父级的，如上所说，可以带来更细粒度的更改区域黑暗模式，刚开始接触scss设计，可能想法不太成熟，请见谅

**Reply by YunYouJun** [reply]
直接挂载根结点过多 css 变量存在性能问题。不知道您指的 scss 设置具体是指？可以具体列一下该部分的代码吗？

**Reply by emojiiii** [reply]
全部css变量都挂载根节点确实会存在性能问题，但是其实并不需要挂载太多的变量，因为我们常用的设置的有且仅用下图这几种，目前根节点设置的有部分变量暂时没有发现在任何地方有使用到，例如 --el-color-#{$type}-light, 如果我们只挂载必要的CSS变量到根节点以便达到方便切换黑暗模式即可
![impicture_20211104_201839](https://user-images.githubusercontent.com/30518686/140312210-699fbfe2-354d-42a0-b50f-af8dba533cb2.png)

**Reply by YunYouJun** [reply]
`--el-color-#{$type}-light` 这部分此前是 button 等使用，以及以便留给用户直接作为变量使用。（现在确实可以考虑移除）
然后 button 内部 class 之所以是具体数值，是因为 scss 是预处理器，只能预先计算出各种混合的变量对应的色阶。
这部分的确有一些别的解决方案，譬如主题内置色彩计算函数，按钮根据色彩之间计算对应的色。
当然这也有一个对应的问题，我们不确定每一个按钮都去计算是否是合理的……

此前自动计算对应按钮色彩一个设想，https://github.com/element-plus/element-plus/pull/3664

**Reply by emojiiii** [reply]
每一个按钮都去计算确实是不太合理的，这会消耗一定的性能
  我希望的是，scss预处理器生成时， .el-button--#{$type} 这里面的可以不用直接生成颜色，而是使用类似这样 --el-button-background-color: var(--el-color-success) 这样的形式，使用根节点的样式，这样在做黑暗模式的时候会不会更加灵活

**Reply by YunYouJun** [reply]
`--el-button-background-color: var(--el-color-success)` 是可能的，这里有一个取舍问题。

譬如，不在按钮内部生成具体数值，则意味着我们需要生成 `--el-color-success`, `--el-color-success-hover`, `--el-color-success-border` 等（以及 warning/danger 等）变量，增加了总体的体积（部分变量对于按需引入可能是不需要的）。当然对于暗黑模式来说则可以省掉 dark 模式下的覆盖样式的体积。

---

我看了下目前其他的 UI 组件库，antdesign 采用的方案是生成了不同类型的 CSS 变量，naiveui 则是为每个按钮生成了内联的变量样式。
这部分可能还需要对比考虑下。我需要实验一下。
如果您有更好的方案，欢迎提出。

**Reply by emojiiii** [reply]
总体的体积不会增加太多，因为只是更改目前已有的样式内的使用，至于增加的:root下的颜色变量，我觉得在默认下的几种类型，并不需要太多额外的颜色变量，加上可以去除现在冗余颜色变量，我认为这是可以值得一试的
更好的方法我觉得是使用时加上:root，假设某一个type一直没被用到，比如danger，其实danger使用到的CSS Variable是不需要挂载到根结点的，按需加载CSS Variable我觉得是一个方向，希望考虑一下，谢谢您的解答和回复

**Reply by YunYouJun** [reply]
关于按需部分的 css 确实是一个很值得探索的方向，目前 WindiCSS 也确实做到了一些类似的效果。
但是由于一直以来的传承（历史架构），element 作为组件库它的 css 是分为了不同组件的 scss 进行预编译，因此没有办法做到更细致的按需加载 css（譬如只使用了 primary 按钮，没有使用 danger 按钮，但是 button 整个样式里仍旧需要 primary danger 等样式）。
（譬如 ant-design 甚至推荐引入所有 css 文件，没有做组件的按需导入。）
而想要做到类 WindiCSS 的按需效果，Element 的样式整个实现方式都需要进行修改，目前暂时还没找到一个合适的方式来进行。

---

WindiCSS 之所以可以做到按需加载，是因为其可以在编译时扫描用户使用的 class 类，直接与用户的源代码建立连接。而对于组件库来说，这一点很难做到，Element 用户可以使用 vite 也可以使用 webpack 等打包工具打包，同时也无法扫描用户用了 button 的哪种类型，从而做到更细粒度的划分。这部分可能还有许多内容需要探索。

**Reply by msidolphin** [reply]
> `--el-button-background-color: var(--el-color-success)` 是可能的，这里有一个取舍问题。
> 
> 譬如，不在按钮内部生成具体数值，则意味着我们需要生成 `--el-color-success`, `--el-color-success-hover`, `--el-color-success-border` 等（以及 warning/danger 等）变量，增加了总体的体积（部分变量对于按需引入可能是不需要的）。当然对于暗黑模式来说则可以省掉 dark 模式下的覆盖样式的体积。
> 
> 我看了下目前其他的 UI 组件库，antdesign 采用的方案是生成了不同类型的 CSS 变量，naiveui 则是为每个按钮生成了内联的变量样式。 这部分可能还需要对比考虑下。我需要实验一下。 如果您有更好的方案，欢迎提出。

我认为naive-ui的实现方案是目前主题切换方面的最佳实践，计算性能几乎可以忽略不计，对于每一个按钮都去计算的问题完全可以通过缓存机制解决

**Comment by YunYouJun** [comment]
Sorry, the questioner used Chinese, so we also used Chinese to quickly explain some of the questions.

Briefly explain what we're talking about above.
We want to simplify the css variable while better adapting to dark themes.

The community currently has some of its own solutions, such as ant-design and naive-ui.
But they all have their own disadvantages, and we're trying to find a balanced solution, but probably no good idea yet.

**Comment by orenmizr** [comment]
@YunYouJun thanks for the responses. 
such great news. the current state of cross component theme-ing is very tricky. 
1. please consider users who need to change your theme completely. what should be the "source of truth" - css3 variables  or scss vars. when they are mixed is very hard to predict the outcome of the changes per component. from experience. 
2. consider using generic variables names... base-background: var(--base-background-color) so it semantic when you change the color variables from light to dark in one root place.
3. don't forget to address about the flash of white background when using dark theme. 
4. i read somewhere you are moving the component variables to their respective files. that can help. you might still want to use design token for cross site (https://theme-ui.com/) like chakra-ui, mui and more.

**Reply by YunYouJun** [reply]
Thanks. With regard to variable names, we will gradually try to be as consistent as possible with tailwindcss.

**Comment by InfoLibre** [comment]
Adding a colourblind theme too? For colorblinded people.

**Reply by YunYouJun** [reply]
@huangzhongzhong Any ideas?

**Reply by orenmizr** [reply]
i think if there is a colour palette to change. as config. you can create whatever theme you want. 

if the infra is good, the rest is self service.

not sure what you mentioned tailwind as resource. it is a utility library. sass is very different use case.

**Reply by YunYouJun** [reply]
I mean the name of tailwind variable. Element's sass variable name is a bit confusing due to historical legacies. For example, `font-color` `text-color` `background-color` `bg-color`, unify them refer to utilities.

**Reply by huangzhongzhong** [reply]
我们有计划提供色弱模式色板，但是可能没那么快哈哈。

**Comment by InfoLibre** [comment]
Are you using Sass in Element+ ? Would you publish a version with Stylus too?

**Reply by InfoLibre** [reply]
https://stylus-lang.com

**Reply by YunYouJun** [reply]
Element UI & Element Plus use sass. We think sass is strong enough and energy constraints that we won't release the stylus version.

**Comment by InfoLibre** [comment]
It's interesting to have a release for Sass, Stylus ans Less, like Fomantic-UI.

https://github.com/fomantic/Fomantic-UI-CSS

https://github.com/fomantic/Fomantic-UI-LESS

https://github.com/fomantic/Fomantic-UI-SASS

https://github.com/fomantic/Fomantic-UI/issues/2152

Developers don't have to learn a new CSS preprocessor syntax for their project.

https://github.com/mojotech/sass2stylus

**Reply by orenmizr** [reply]
i am not a part of the element plus team. 
but building infrastructure that support using each artifacts for different kind of style builds will make their library more complex and i don't think they would want to do it.

**Comment by orenmizr** [comment]
@YunYouJun does anyone at ElementPlus  know about a theme refactor release ? dark/light or just fix the styling css3/saas config ?

**Reply by YunYouJun** [reply]
We are currently resizing the size. It's also a break change. It will meet you in the next two weeks.

Dark mode is certainly inevitable, but I'm sorry we have not yet had time to fully engage.

**Reply by bypanghu** [reply]
Sorry to disturb you. I would like to know when the dark mode will be available online? One of our open source projects may need to be upgraded and adjusted in the near future. It would be better if the dark mode is upgraded.So, can you give me an approximate time to go online?

**Reply by YunYouJun** [reply]
We plan to release the feature in the first quarter of next year.

**Reply by bypanghu** [reply]
thankyou !

**Comment by huangzhongzhong** [comment]
上线时间预计在 2022年2月份或者更晚

**Reply by bypanghu** [reply]
谢谢

**Comment by tolking** [comment]
Suggestions on implementing theme configuration, modification, nesting, and customization through `el-config-provider`

关于通过 `el-config-provider` 实现主题配置、修改、嵌套、自定义的建议

## Api

theme

- Type: 'light' | 'dark' | 'auto' | CSSProperties | string
- Default: 'light'

## Use

``` vue
<el-config-provider :theme="theme">
  <!-- Components  -->
</el-config-provider>
```

### light

theme="light" -> class="el-theme-light" (or theme="light")

```css
:root,
.el-theme-light { /* (or [theme=light]) */
  /* css-variables of light */
}
```

### dark

theme="dark" -> class="el-theme-dark"

```css
.el-theme-dark {
  /* css-variables of dark */
}
```

### auto

theme="auto" -> class="el-theme-auto"

```css
@media (prefers-color-scheme: dark) {
  .el-theme-auto {
    /* css-variables of dark */
  }
}
```

### CSSProperties

:theme="{ '--el-bg-color': '#f00' }" -> style="--el-bg-color:#f00"

### Custom theme

theme="my-theme" -> class="el-theme-my-theme"

add css file

```css
.el-theme-my-theme {
  /* css-variables of custom theme */
}
```

**Reply by YunYouJun** [reply]
Thanks!

**Comment by baermathias** [comment]
Is this feature already available?

**Reply by YunYouJun** [reply]
Sorry, its design is not yet finished.

**Comment by netangsoft** [comment]
@YunYouJun 请问这次暗黑模式和即将要发布的稳定版会一起发布吗？

**Reply by sxzz** [reply]
不会，还没做。

**Comment by TothingWay** [comment]
How is the progress of the dark theme?

**Reply by sxzz** [reply]
https://github.com/element-plus/element-plus/issues/5712

**Reply by YunYouJun** [reply]
Designer is adjusting the palette.

**Comment by ZHT131** [comment]
此功能预计多久能上线呢？

**Reply by YunYouJun** [reply]
The feature is almost ready and we're polishing the details. It will be online this month.

**Reply by ZHT131** [reply]
> 该功能几乎准备就绪，我们正在完善细节。本月将上线。

加油，终于快等到了

**Comment by orenmizr** [comment]
what is the expected change ? can the team give a bit more details on the change ?  🙏

**Reply by YunYouJun** [reply]
What does expected change mean?

**Reply by orenmizr** [reply]
how will it look like ?
where will the styles sit ? 
will it be scss or css variables ? 
will it be one object or per component ? 
how will we override the default ?

**Reply by sxzz** [reply]
You can preview docs here 
https://staging.element-plus.org/en-US/component/button.html?feature:theme=true

**Reply by orenmizr** [reply]
thanks!
```
<el-button color="#626aef" :dark="isDark">Default</el-button>
```

looks like dark is controlled via the component.. not some global settings. is that true ? or just a an example.

**Reply by sxzz** [reply]
Just an example. The option will be provided by `config-provider`

**Reply by orenmizr** [reply]
thanks!

**Reply by YunYouJun** [reply]
The default button will automatically follow the class `.dark`, and the custom color button is a bit special because it's runtime and allows you to actively set its light and dark modes.

**Reply by YunYouJun** [reply]
We handle all component styling variables by css vars and individually adapted the details of each component, so we can quickly enable dark mode by importing an scss or compiled css file.

The end goal might look like this:

```ts
// main.ts
// Of course, compiled css is provided too
import 'element-plus/theme-chalk/src/dark/css-vars.scss'
```

**Reply by orenmizr** [reply]
the current version was very hard to override... i got scss and css variables clashing... often changing just parts of a component and not all. 
whatever the final solution be - i hope all changeable variables will be in one place per component so you can gain granular control of every aspect of its style.

**Reply by YunYouJun** [reply]
You are right. We organized the dark variables in a scss file.

**Comment by InfoLibre** [comment]
Will you publish Stylus files too?

**Reply by YunYouJun** [reply]
I don't think we will, because we use sass's recent modules and forwarding features, and have no energy and plans to adapt to stylus.

**Comment by NO-MAP** [comment]
你好，DARK MODE 预计什么时候发布呢？

**Reply by NO-MAP** [reply]
> 你好，DARK MODE 预计什么时候发布呢？

哈喽，4月份能把这个功能发布出来吗，哈喽哈喽。

**Reply by sxzz** [reply]
> > 你好，DARK MODE 预计什么时候发布呢？
> 
> 
> 
> 哈喽，4月份能把这个功能发布出来吗，哈喽哈喽。

劳动节假期后
/cc @huangzhongzhong

**Reply by orenmizr** [reply]
is this correct? release expected "after labor day" 5th of september ?

**Reply by NO-MAP** [reply]
> 

after May 6th

**Comment by sxzz** [comment]
Version 2.2 is expected to be released around May 8th. With features:
- Virtualized table
- Dark mode theme

**Reply by xuwenhao1995** [reply]
5月9号啦，求深色主题😊

**Reply by YunYouJun** [reply]
Of course, dark mode has been released in `v2.2.0`.

**Comment by imzbf** [comment]
坐等😀

**Comment by NO-MAP** [comment]
坐等 +1

**Comment by YunYouJun** [comment]
Dark mode has been released in `v2.2.0`. 

Since this is the first release for dark mode, there may still be some details that are not perfect.
Feedback is always welcome, we will make it better as soon as possible.

More info see https://element-plus.org/zh-CN/guide/dark-mode.html .

**Reply by orenmizr** [reply]
that is absolutely amazing.... can't wait to play with. 🔥

**Comment by orenmizr** [comment]
what is the correct way to change elements - i assume it is using css vars
for the menu - color: getCssVar('**menu-text-color**'); I didnt see it in **menu.scss** or it is written in a way i did not recognize, 

1. where can i find that value ? 
2. should i extend/override it in the theme ? what is the recommendation

**Reply by YunYouJun** [reply]
You can find all var in [packages/theme-chalk/src/common/var.scss](https://github.com/element-plus/element-plus/blob/dev/packages/theme-chalk/src/common/var.scss).

You can still define variables via scss, they will be automatically generated as css vars.

We've been hoping for a script to generate documentation, but unfortunately there's no progress.
As a temporary, I think accessing [packages/theme-chalk/src/common/var.scss](https://github.com/element-plus/element-plus/blob/dev/packages/theme-chalk/src/common/var.scss) directly can also solve your problems.

**Reply by orenmizr** [reply]
i will give more info for my question: 
getCssVar('**menu-text-color**') means that someone **had to a set for that color.**  but it is not in var.scss or anywhere i could see https://github.com/element-plus/element-plus/search?q=text-color-regular

**Reply by tolking** [reply]
The [$text-color-regular](https://github.com/element-plus/element-plus/blob/dev/packages/theme-chalk/src/common/var.scss#L100)  for scss users. 

The [el-var](https://unpkg.com/browse/element-plus@2.2.0/theme-chalk/el-var.css) and [dark/css-vars](https://unpkg.com/browse/element-plus@2.2.0/theme-chalk/dark/css-vars.css) for css users.

Override the corresponding `var` according to the situation.

**Reply by YunYouJun** [reply]
```scss
$text-color: () !default;
$text-color: map.merge(
  (
    'primary': #303133,
    'regular': #606266,
    'secondary': #909399,
    'placeholder': #a8abb2,
    'disabled': #c0c4cc,
  ),
  $text-color
);
```

Will be generated to `--el-text-color-regular` and others.

You can override the css variable or directly modify the scss to change the initial generated value.

**Comment by hooray** [comment]
我想同时修改默认和暗黑模式下的 $bg-color 变量
```scss
@forward "element-plus/theme-chalk/src/common/var.scss" with (
    $bg-color: (
        "": #f2f3f5,
        "page": #ffffff,
        "overlay": #ffffff,
    )
);
@use "element-plus/theme-chalk/src/index.scss" as *;
@forward "element-plus/theme-chalk/src/dark/var.scss" with (
    $bg-color: (
        "page": #312d4b,
        "": #28243d,
        "overlay": #1d1e1f,
    )
);
@use "element-plus/theme-chalk/src/dark/css-vars.scss" as *;
```

但是运行有报错，删除任何一段均可生效，但同时放到一起就不行

![image](https://user-images.githubusercontent.com/1574903/168948432-f5bf731c-bae4-4828-81db-2e1baf8b4e44.png)

**Reply by hooray** [reply]
当我把这两段拆分到 2 个文件里去 @import 就不会有问题

**Reply by YunYouJun** [reply]
Since both files contain the same variable `$colors`, we need different files to resolve the namespace issue.
And since `@import` will eventually be deprecated, I recommend always using the `@use` syntax.

**Reply by laudaikinhdi** [reply]
I dont still understand. Can you an example code?

**Reply by YunYouJun** [reply]
OK. I create a example in https://github.com/element-plus/element-plus-vite-starter.

Switch between light and dark, and you can find that the theme color of the button is different.
Hope this helps you.

**Reply by laudaikinhdi** [reply]
Hi @YunYouJun 
Can you help me with this issue?
I cannot custom theme with mode "light" but mode "dark" is working
![image](https://user-images.githubusercontent.com/34116228/172588379-9d1eacdf-d4bb-4c92-9355-2f92e7fcaa2a.png)
webpack.config.js (laravel-mix)
![image](https://user-images.githubusercontent.com/34116228/172588455-85b595cc-c95c-4ff9-92b6-3efb5ca5e70e.png)

**Reply by YunYouJun** [reply]
You need to customize dark in another file, please refer to the example I mentioned above, I think it has solved this thing.

**Reply by laudaikinhdi** [reply]
> You need to customize dark in another file, please refer to the example I mentioned above, I think it has solved this thing.

You do not understand my mind. The example its working with "dark" mode, but "light" mode I cannot custom theme

**Reply by YunYouJun** [reply]
<img width="1561" alt="image" src="https://user-images.githubusercontent.com/25154432/172653943-503e5402-8386-49b0-b7b6-832975cb889f.png">

The effect of customizing the light theme has been shown.

**Reply by laudaikinhdi** [reply]
I see. But I using webpack v5 build then light mode cannot custom theme

**Reply by YunYouJun** [reply]
Please create an issue and provide a reproducible repository.

**Reply by laudaikinhdi** [reply]
You can example with webpack?

**Reply by orenmizr** [reply]
i love element plus... but i tried to custom dark theme the grid and could not make it look as designed... the css variables + sass + sass components + index's css import... its too vague not sure what to do

**Reply by YunYouJun** [reply]
@orenmizr Try this example? https://github.com/element-plus/element-plus-vite-starter

If you have a specific question, you can create an issue and at me.

**Comment by ryuhangyeong** [comment]
awesome!

**Comment by Red-Asuka** [comment]
How can I inherit and extend a theme? For example, besides the light and dark themes, I want a night theme that inherits from the dark theme and can be freely extended.

**Reply by Dsaquel** [reply]
Maybe this [repo ](https://github.com/element-plus/colors) from element-plus is what you need
